component/src/main/java/org/wso2/extension/siddhi/io/http/source/HttpConnectorRegistry.java

```diff
@@ -27,25 +27,144 @@
 import org.wso2.extension.siddhi.io.http.source.util.HttpSourceUtil;
+import org.wso2.siddhi.core.exception.ExecutionPlanCreationException;
+import org.wso2.siddhi.core.stream.input.source.SourceEventListener;
 import org.wso2.siddhi.core.util.config.ConfigReader;
 
-import java.util.Collections;
-import java.util.HashSet;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
 /**
  * {@code HttpConnectorRegistry} The code is responsible for maintaining the all active server connectors.
  */
 class HttpConnectorRegistry {
-    private static final Logger log = Logger.getLogger(HttpConnectorRegistry.class);
     private static HttpConnectorRegistry instance = new HttpConnectorRegistry();
-    private final Map<String, HTTPServerConnector> registeredServerConnectors = new ConcurrentHashMap<>();
+    private final Logger log = Logger.getLogger(HttpConnectorRegistry.class);
+    private final Map<String, HTTPServerConnector> serverConnectorMap = new ConcurrentHashMap<>();
+    private final Map<String, HttpSourceListener> sourceListenersMap = new ConcurrentHashMap<>();
     private HttpMessageProcessor httpMessageProcessor = new HttpMessageProcessor();
+    private ServerConnectorController serverConnectorController;
 
-    private HttpConnectorRegistry() {
+    private HttpConnectorRegistry() {}
+
+    /**
+     * Get the source listener map.
+     *
+     * @return the source listener map
+     */
+    Map<String, HttpSourceListener> getSourceListenersMap() {
+        return this.sourceListenersMap;
+    }
+
+    /**
+     * Initialize and start the server connector controller.
+     *
+     * @param sourceConfigReader the siddhi source config reader.
+     */
+    void initHttpServerConnector(ConfigReader sourceConfigReader) {
+        if (this.serverConnectorController == null) {
+            TransportsConfiguration configuration = new TransportsConfiguration();
+            configuration.setTransportProperties(new HttpSourceUtil().getTransportConfigurations
+                    (sourceConfigReader));
+            this.serverConnectorController = new ServerConnectorController(configuration);
+            this.serverConnectorController.start();
+        }
+    }
+
+    /**
+     * Register new source listener.
+     *
+     * @param sourceEventListener the source event listener.
+     * @param listenerUrl the listener url.
+     * @param workerThread the worker thread count of siddhi level thread pool executor.
+     * @param isAuth the authentication is required for source listener.
+     */
+    void registerSourceListener(SourceEventListener sourceEventListener, String listenerUrl, int
+            workerThread, Boolean isAuth) {
+        synchronized (this.sourceListenersMap) {
+            if (this.sourceListenersMap.get(HttpSourceUtil.getSourceListenerKey(listenerUrl)) != null) {
+                throw new ExecutionPlanCreationException("Listener URL " + listenerUrl + " already connected.");
```

Comment:
HttpSourceUtil.getSourceListenerKey(listenerUrl) is called twice. use local variables and make only one call